File: jaws-spark-sql-data/src/main/java/com/xpatterns/jaws/data/impl/JawsHdfsLogging.java
//package com.xpatterns.jaws.data.impl;
//
//import java.io.IOException;
//import java.util.ArrayList;
//import java.util.Collection;
//import java.util.Comparator;
//import java.util.Iterator;
//import java.util.LinkedList;
//import java.util.SortedSet;
//
//import org.apache.hadoop.conf.Configuration;
//import org.apache.log4j.Logger;
//
//import com.xpatterns.jaws.data.DTO.LogDTO;
//import com.xpatterns.jaws.data.DTO.ScriptMetaDTO;
//import com.xpatterns.jaws.data.DTO.StateDTO;
//import com.xpatterns.jaws.data.contracts.IJawsLogging;
//import com.xpatterns.jaws.data.utils.QueryState;
//import com.xpatterns.jaws.data.utils.Utils;
//
//public class JawsHdfsLogging implements IJawsLogging {
//
//	Configuration configuration;
//	public static final String QUERYID_SEPARATOR = "-----";
//
//	private static Logger logger = Logger.getLogger(JawsHdfsLogging.class.getName());
//
//	public JawsHdfsLogging() {
//
//	}
//
//	public JawsHdfsLogging(Configuration configuration) throws Exception {
//		// getting the properties from the properties file
//
//		this.configuration = configuration;
//		boolean forcedMode = configuration.getBoolean(Utils.FORCED_MODE, false);
//
//		Utils.createFolderIfDoesntExist(configuration, configuration.get(Utils.LOGGING_FOLDER), forcedMode);
//		Utils.createFolderIfDoesntExist(configuration, configuration.get(Utils.STATUS_FOLDER), forcedMode);
//		Utils.createFolderIfDoesntExist(configuration, configuration.get(Utils.DETAILS_FOLDER), forcedMode);
//		Utils.createFolderIfDoesntExist(configuration, configuration.get(Utils.METAINFO_FOLDER), forcedMode);
//	}
//
//	@Override
//	public void setState(String uuid, QueryState type) throws Exception {
//
//		logger.debug("Writing query state " + type.toString() + " to query " + uuid + " on hdfs");
//		Utils.rewriteFile(type.name(), configuration, configuration.get(Utils.STATUS_FOLDER) + "/" + uuid);
//
//	}
//
//	@Override
//	public void setScriptDetails(String uuid, String scriptDetails) throws Exception {
//
//		logger.info("Writing script details " + scriptDetails + " to query " + uuid);
//		Utils.rewriteFile(scriptDetails, configuration, configuration.get(Utils.DETAILS_FOLDER) + "/" + uuid);
//
//	}
//
//	@Override
//	public void addLog(String uuid, String queryId, Long time, String log) throws Exception {
//
//		logger.debug("Writing log " + log + " to query " + uuid + " at time " + time);
//		String folderName = configuration.get(Utils.LOGGING_FOLDER) + "/" + uuid;
//		String fileName = folderName + "/" + time.toString();
//		String logMessage = queryId + QUERYID_SEPARATOR + log;
//		Utils.createFolderIfDoesntExist(configuration, folderName, false);
//		Utils.rewriteFile(logMessage, configuration, fileName);
//
//	}
//
//	@Override
//	public QueryState getState(String uuid) throws IOException {
//
//		logger.info("Reading query state for query: " + uuid);
//		String state = Utils.readFile(configuration, configuration.get(Utils.STATUS_FOLDER) + "/" + uuid);
//		return QueryState.valueOf(state);
//	}
//
//	@Override
//	public String getScriptDetails(String uuid) throws IOException {
//
//		logger.info("Reading script details for query: " + uuid);
//		return Utils.readFile(configuration, configuration.get(Utils.DETAILS_FOLDER) + "/" + uuid);
//
//	}
//
//	@Override
//	public Collection<LogDTO> getLogs(String uuid, Long time, int limit) throws IOException {
//
//		logger.info("Reading logs for query: " + uuid + " from date: " + time);
//		Collection<LogDTO> logs = new ArrayList<LogDTO>();
//
//		String folderName = configuration.get(Utils.LOGGING_FOLDER) + "/" + uuid;
//
//		SortedSet<String> files = Utils.listFiles(configuration, folderName, new Comparator<String>() {
//
//			@Override
//			public int compare(String o1, String o2) {
//				return o1.compareTo(o2);
//			}
//
//		});
//
//		if (files.contains(time.toString())) {
//			files = files.tailSet(time.toString());
//		}
//
//		Collection<String> filesToBeRead = getSubset(limit, files);
//
//		for (String file : filesToBeRead) {
//			String[] logedInfo = Utils.readFile(configuration, folderName + "/" + file).split(QUERYID_SEPARATOR);
//			if (logedInfo.length == 2) {
//				logs.add(new LogDTO(logedInfo[1], logedInfo[0], Long.parseLong(file)));
//			}
//
//		}
//
//		return logs;
//	}
//
//	private Collection<String> getSubset(int limit, SortedSet<String> files) {
//		Collection<String> filesToBeRead = new LinkedList<String>();
//
//		Iterator<String> iterator = files.iterator();
//		while (iterator.hasNext() && limit > 0) {
//			String file = iterator.next();
//			filesToBeRead.add(file);
//			limit--;
//		}
//
//		return filesToBeRead;
//	}
//
//	@Override
//	public Collection<StateDTO> getQueriesStates(String uuid, int limit) throws IOException {
//
//		logger.info("Reading states for queries starting with the query: " + uuid);
//		Collection<StateDTO> stateList = new ArrayList<StateDTO>();
//
//		String folderName = configuration.get(Utils.STATUS_FOLDER);
//		String startFilename = folderName + "/" + uuid;
//		SortedSet<String> files = Utils.listFiles(configuration, folderName, new Comparator<String>() {
//
//			@Override
//			public int compare(String o1, String o2) {
//				return o2.compareTo(o1);
//			}
//
//		});
//
//		if (files.contains(startFilename)) {
//			files = files.tailSet(startFilename);
//		}
//
//		Collection<String> filesToBeRead = getSubset(limit, files);
//
//		for (String file : filesToBeRead) {
//			stateList.add(new StateDTO(QueryState.valueOf(Utils.readFile(configuration, folderName + "/" + file)), Utils.getNameFromPath(file)));
//
//		}
//
//		return stateList;
//	}
//
//	@Override
//	public void setMetaInfo(String uuid, ScriptMetaDTO metainfo) throws Exception {
//		logger.info("Writing script metainfo " + metainfo + " to query " + uuid);
//		String buffer = metainfo.toJson();
//		Utils.rewriteFile(buffer.getBytes(), configuration, configuration.get(Utils.METAINFO_FOLDER) + "/" + uuid);
//
//	}
//
//	@Override
//	public ScriptMetaDTO getMetaInfo(String uuid) throws IOException {
//		logger.info("Reading query metainfo for query: " + uuid);
//		byte[] bytes = Utils.readBytesFromFile(configuration, configuration.get(Utils.METAINFO_FOLDER) + "/" + uuid);
//		ScriptMetaDTO result = ScriptMetaDTO.fromJson(new String(bytes));
//
//		return result;
//	}
//
//}

File: jaws-spark-sql-data/src/main/scala/com/xpatterns/jaws/data/impl/JawsHdfsLogging.scala
package com.xpatterns.jaws.data.impl

import org.apache.hadoop.conf.Configuration
import org.apache.log4j.Logger
import com.xpatterns.jaws.data.contracts.TJawsLogging
import com.xpatterns.jaws.data.DTO.Logs
import java.util.Comparator
import java.util.SortedSet
import com.xpatterns.jaws.data.DTO.Log
import com.xpatterns.jaws.data.DTO.Queries
import com.xpatterns.jaws.data.DTO.Query
import com.xpatterns.jaws.data.DTO.QueryMetaInfo
import net.liftweb.json._
import spray.json._
import com.xpatterns.jaws.data.utils.QueryState
import com.xpatterns.jaws.data.utils.Utils
import org.apache.hadoop.fs.Path
import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.io.IOUtils

class JawsHdfsLogging(configuration: Configuration) extends TJawsLogging {

  val QUERYID_SEPARATOR = "-----"

  val logger = Logger.getLogger("JawsHdfsLogging")

  val forcedMode = configuration.getBoolean(Utils.FORCED_MODE, false)
  Utils.createFolderIfDoesntExist(configuration, configuration.get(Utils.LOGGING_FOLDER), forcedMode)
  Utils.createFolderIfDoesntExist(configuration, configuration.get(Utils.STATUS_FOLDER), forcedMode)
  Utils.createFolderIfDoesntExist(configuration, configuration.get(Utils.DETAILS_FOLDER), forcedMode)
  Utils.createFolderIfDoesntExist(configuration, configuration.get(Utils.METAINFO_FOLDER), forcedMode)

  override def setState(uuid: String, queryState: QueryState.QueryState) {

    logger.debug("Writing query state " + queryState.toString() + " to query " + uuid)
    Utils.rewriteFile(queryState.toString(), configuration, configuration.get(Utils.STATUS_FOLDER) + "/" + uuid)

  }

  override def setScriptDetails(queryId: String, scriptDetails: String) {

    logger.debug("Writing script details " + scriptDetails + " to query " + queryId)
    Utils.rewriteFile(scriptDetails, configuration, configuration.get(Utils.DETAILS_FOLDER) + "/" + queryId)

  }

  override def addLog(queryId: String, jobId: String, time: Long, log: String) {

    logger.debug("Writing log " + log + " to query " + queryId + " at time " + time)

    logger.debug("Writing log " + log + " to query " + queryId + " at time " + time)
    val folderName = configuration.get(Utils.LOGGING_FOLDER) + "/" + queryId
    val fileName = folderName + "/" + time.toString()
    val logMessage = jobId + QUERYID_SEPARATOR + log
    Utils.createFolderIfDoesntExist(configuration, folderName, false)
    Utils.rewriteFile(logMessage, configuration, fileName)

  }

  override def getState(queryId: String): QueryState.QueryState = {

    logger.debug("Reading query state for query: " + queryId)
    val filename = configuration.get(Utils.STATUS_FOLDER) + "/" + queryId

    if (Utils.checkFileExistence(filename, configuration)) {
      val state = Utils.readFile(configuration, filename)
      return QueryState.withName(state)
    }
    return QueryState.NOT_FOUND
  }

  override def getScriptDetails(queryId: String): String = {
    logger.info("Reading script details for query: " + queryId)
    val filename = configuration.get(Utils.DETAILS_FOLDER) + "/" + queryId
    return if (Utils.checkFileExistence(filename, configuration)) Utils.readFile(configuration, filename) else ""

  }

  override def getLogs(queryId: String, time: Long, limit: Int): Logs = {

    logger.debug("Reading logs for query: " + queryId + " from date: " + time)

    val state = getState(queryId).toString
    val folderName = configuration.get(Utils.LOGGING_FOLDER) + "/" + queryId
    var logs = Array[Log]()
    var files = Utils.listFiles(configuration, folderName, new Comparator[String]() {

      override def compare(o1: String, o2: String): Int = {
        return o1.compareTo(o2)
      }

    })

    if (files.contains(time.toString())) {
      files = files.tailSet(time.toString())
    }

    val filesToBeRead = getSubset(limit, files)

    filesToBeRead.foreach(file => {
      val logedInfo = Utils.readFile(configuration, folderName + "/" + file).split(QUERYID_SEPARATOR)
      if (logedInfo.length == 2) {
        logs = logs ++ Array(new Log(logedInfo(1), logedInfo(0), file.toLong))

      }
    })

    return new Logs(logs, state)
  }

  def getSubset(limit: Int, files: SortedSet[String]): List[String] = {
    var filesToBeRead = List[String]()
    var limitMutable = limit

    val iterator = files.iterator()
   
    while (iterator.hasNext() && limitMutable > 0) {
      val file = iterator.next()
      filesToBeRead = filesToBeRead ++ List(file)
      limitMutable = limitMutable - 1
    }

    return filesToBeRead
  }

  override def getQueriesStates(queryId: String, limit: Int): Queries = {
    val queryIdValue = Option(queryId).getOrElse("")
    logger.info("Reading states for queries starting with the query: " + queryIdValue)
    var stateList = Array[Query]()

    val folderName = configuration.get(Utils.STATUS_FOLDER)
    val startFilename = folderName + "/" + queryIdValue
    var files = Utils.listFiles(configuration, folderName, new Comparator[String]() {

      override def compare(o1: String, o2: String): Int = {
        return o2.compareTo(o1)
      }

    })

    if (files.contains(queryIdValue)) {
      files = files.tailSet(queryIdValue)
      files.remove(queryIdValue)
    }

    val filesToBeRead = getSubset(limit, files)

    filesToBeRead.foreach(file => {
      val currentUuid = Utils.getNameFromPath(file)
      stateList = stateList ++ Array(new Query(Utils.readFile(configuration, folderName + "/" + file), currentUuid, getScriptDetails(currentUuid)))
    })

    return new Queries(stateList)
  }

  override def setMetaInfo(queryId: String, metainfo: QueryMetaInfo) {
    logger.debug("Writing script meta info " + metainfo + " to query " + queryId)
    val buffer = metainfo.toJson.toString
    Utils.rewriteFile(buffer, configuration, configuration.get(Utils.METAINFO_FOLDER) + "/" + queryId)
  }

  override def getMetaInfo(queryId: String): QueryMetaInfo = {
    logger.debug("Reading meta info for for query: " + queryId)

    val value = Utils.readFile(configuration, configuration.get(Utils.METAINFO_FOLDER) + "/" + queryId)

    implicit val formats = DefaultFormats

    val json = parse(value)
    return json.extract[QueryMetaInfo]

  }
  
  def deleteQuery(queryId: String) {}
}
